### ISSUE TITLE
Provide a way to know how many lines remain in the node currently running

### ISSUE THREAD

**Issue by satanas** [YarnSpinnerTool/YarnSpinner#237]
**Is your feature request related to a problem? Please describe.**

I'm designing a dialogue interface that includes a blinking icon in the bottom-right corner to hint the user that they must click/tap to continue and I'd like this icon to be an arrow when there are still lines of dialogue and a square when we have reached the end of the node. However, that's not possible right now because there are no ways (or at least not that I know of) to know how many lines are left in the current node or if we reached the end of it (before the `DialogueComplete` signal).

![dialog-interface](https://user-images.githubusercontent.com/29679/81121092-ce1d4800-8ee2-11ea-92ea-80b5c4e836c1.gif)
The image above should illustrate the dialogue interface I'm talking about.

**Describe the solution you'd like**
Maybe the `Line` class received in the `RunLine(Line line, IDictionary<string, string> strings, Action onLineComplete)` method could have a bool param (like `isEnd`, `endOfNode`, etc) to indicate if we have reached the last dialogue line or not.

**Describe alternatives you've considered**
The `DialogueComplete` signal doesn't work because I receive it only after the player has clicked the last line of dialogue already. Instead, I use that method to close the interface when everything is done.

**Additional context**
N/A

**Comment by McJones** [comment]
Assuming I am correctly understanding what you are after you want this is actually a really tricky problem because we can't realistically know how many lines are left in a node until we finish the node.
We can make a guess based on the total number of lines in the node but until we actually hit a jump into another node or run out of lines we can't tell.
As an example of this:

```
<<if $var is true>>
  Oh, awesome.
  thanks so much
<<else>>
  Oh, ok
<<endif>>
```

In this small snippet are there two or one lines left?
In this small case it is pretty trivial to work it out, but solving it becomes a lot harder as soon as more variables get involved (and if you add in loops or recursion it is known to be impossible).

Some potential fixes:
- could you make it so that the user clicking again closes the UI after receiving the `DialogueComplete` signal instead of automatically closing it?
- add a `<<last line>>` command above the last line in your dialogue that lets you know when there is only one line left
- Dive into the stack of instructions inside the virtual machine and count off the number of run line instructions remaining before you hit the end or a node jump.

Personally I would go with the first option because the second option is just asking for bugs to be introduced later on, and the third options is just gonna be some nasty code to write.
I'm going to leave this issue open for now, please let me know if I've understood you correctly or if I have just gone off on a rant about something unrelated.

**Comment by McJones** [comment]
There's been no activity on this issue for over a week now, I am closing it.
Feel free to reopen it or open a new issue if you have a similar concern.

**Comment by satanas** [comment]
Hey @McJones, sorry for my delayed response and thanks for clarifying the issue. I see this is not something that can be implemented easily and your second suggestion makes a lot sense. Actually, I think I can implement something with a queue to buffer some of the instructions there and count them before actually processing them. Thanks for the suggestion!

**Comment by daerogami** [comment]
@McJones So in the same vein, there isn't a way to know that after getting a line that options will be presented? I have to effectively prompt for a redundant input to show options or buffer lines as @satanas suggests?

**Comment by desplesda** [comment]
I can think of one simple-enough way to determine this, which might result in false negatives but would never result in a false-negative: scan through the instructions on the current node, starting from the current position, until you hit either a flow-control instruction or an instruction to present options. If the instruction you stopped at was an instruction to present options, then you know that options are guaranteed to be coming up. (The inverse isn't necessarily true, though.)

**Comment by daerogami** [comment]
@desplesda How are you proposing I scan through the current node? I don't see any way to interface with the contents of a node via the Dialogue object.

I tried digging into the program object but it has some odd type for `program.Nodes` (MapField<string,Node>) which requires Google.protobuf to get intellisense on. That seems deeper under-the-hood than should be necessary.

**Comment by desplesda** [comment]
Apologies, I didn't mean to imply that this was a feature that's easily implementable using only the existing public APIs, but rather it's a sketch of an idea for how the question could be figured out at all.

**Comment by daerogami** [comment]
No, my apologies for coming off as indignant. Just seeking clarification to ensure there wasn't something I was missing.

**Comment by daerogami** [comment]
Dusted off my project a few days ago. I found a way to do this.
I implemented a simple command called `noprompt`
This tells my dialogue handler for lines to continue execution after displaying the following line.

Sample yarn node:
```
Welcome to the noprompt example!
<<noprompt>>
You should see options below without having to click after displaying this text.
[[Go to Yes|Yes]]
[[Go to No|No]]
```

dialogue handlers
```
private HandlerExecutionType HandleLine(Line line)
{
    _strings.TryGetValue(line.ID, out var info);
    TextLabel.Text = info.text;
    if (_autoAcknowledgeNextLine)
    {
        return HandlerExecutionType.ContinueExecution;
    }
    _awaitingAcknowledgement = true;
    return HandlerExecutionType.PauseExecution;
}

private HandlerExecutionType HandleCommand(Command command)
{
    if (command.Text == "noprompt")
    {
        _autoAcknowledgeNextLine = true;
    }
    return HandlerExecutionType.ContinueExecution;
}
```
